File: src/org/jetbrains/r/highlighting/RHighlighter.java
// Copyright (c) 2017, Holger Brandl, Ekaterina Tuzova
/*
 * Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
 */

package org.jetbrains.r.highlighting;

import com.intellij.lexer.Lexer;
import com.intellij.openapi.editor.colors.TextAttributesKey;
import com.intellij.openapi.fileTypes.SyntaxHighlighterBase;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.r.lexer.RLexer;
import org.jetbrains.r.parsing.RElementTypes;
import org.jetbrains.r.parsing.RParserDefinition;
import org.jetbrains.r.psi.RPsiUtil;

import java.util.HashMap;
import java.util.Map;

public class RHighlighter extends SyntaxHighlighterBase {
    private static final Map<IElementType, TextAttributesKey> ATTRIBUTES = new HashMap<IElementType, TextAttributesKey>();

    static {
        fillMap(ATTRIBUTES, RPsiUtil.INSTANCE.getRESERVED_WORDS(), RHighlighterColors.Companion.getKEYWORD());
        fillMap(ATTRIBUTES, RPsiUtil.INSTANCE.getOPERATORS(), RHighlighterColors.Companion.getOPERATION_SIGN());

        ATTRIBUTES.put(RElementTypes.R_STRING, RHighlighterColors.Companion.getSTRING());
        ATTRIBUTES.put(RElementTypes.R_NUMERIC, RHighlighterColors.Companion.getNUMBER());
        ATTRIBUTES.put(RElementTypes.R_COMPLEX, RHighlighterColors.Companion.getNUMBER());
        ATTRIBUTES.put(RElementTypes.R_INTEGER, RHighlighterColors.Companion.getNUMBER());


        ATTRIBUTES.put(RElementTypes.R_LPAR, RHighlighterColors.Companion.getPARENTHESES());
        ATTRIBUTES.put(RElementTypes.R_RPAR, RHighlighterColors.Companion.getPARENTHESES());

        ATTRIBUTES.put(RElementTypes.R_LBRACE, RHighlighterColors.Companion.getBRACES());
        ATTRIBUTES.put(RElementTypes.R_RBRACE, RHighlighterColors.Companion.getBRACES());

        ATTRIBUTES.put(RElementTypes.R_LBRACKET, RHighlighterColors.Companion.getBRACKETS());
        ATTRIBUTES.put(RElementTypes.R_LDBRACKET, RHighlighterColors.Companion.getBRACKETS());
        ATTRIBUTES.put(RElementTypes.R_RBRACKET, RHighlighterColors.Companion.getBRACKETS());
        ATTRIBUTES.put(RElementTypes.R_RDBRACKET, RHighlighterColors.Companion.getBRACKETS());

        ATTRIBUTES.put(RElementTypes.R_COMMA, RHighlighterColors.Companion.getCOMMA());
        ATTRIBUTES.put(RElementTypes.R_SEMI, RHighlighterColors.Companion.getSEMICOLON());

        ATTRIBUTES.put(RParserDefinition.END_OF_LINE_COMMENT, RHighlighterColors.Companion.getLINE_COMMENT());

        ATTRIBUTES.put(RParserDefinition.BAD_CHARACTER, RHighlighterColors.Companion.getBAD_CHARACTER());
    }

    @Override
    @NotNull
    public Lexer getHighlightingLexer() {
        return new RLexer();
    }


    @Override
    @NotNull
    public TextAttributesKey[] getTokenHighlights(IElementType tokenType) {
        return pack(ATTRIBUTES.get(tokenType));
    }
}

File: src/org/jetbrains/r/highlighting/RHighlighter.kt
// Copyright (c) 2017, Holger Brandl, Ekaterina Tuzova
/*
 * Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
 */

package org.jetbrains.r.highlighting

import com.intellij.lexer.Lexer
import com.intellij.openapi.editor.colors.TextAttributesKey
import com.intellij.openapi.fileTypes.SyntaxHighlighterBase
import com.intellij.psi.tree.IElementType
import com.intellij.r.psi.lexer.RLexer
import com.intellij.r.psi.parsing.RElementTypes
import com.intellij.r.psi.parsing.RTokenTypes
import com.intellij.r.psi.psi.RPsiUtil

class RHighlighter : SyntaxHighlighterBase() {

  override fun getHighlightingLexer(): Lexer {
    return RLexer()
  }

  override fun getTokenHighlights(tokenType: IElementType): Array<TextAttributesKey> {
    return pack(ATTRIBUTES[tokenType])
  }

  companion object {
    private val ATTRIBUTES = HashMap<IElementType, TextAttributesKey>()

    init {
      fillMap(ATTRIBUTES, RPsiUtil.RESERVED_WORDS, KEYWORD)
      fillMap(ATTRIBUTES, RPsiUtil.OPERATORS, OPERATION_SIGN)

      ATTRIBUTES[RElementTypes.R_STRING] = STRING
      ATTRIBUTES[RElementTypes.R_INVALID_STRING] = STRING
      ATTRIBUTES[RElementTypes.R_NUMERIC] = NUMBER
      ATTRIBUTES[RElementTypes.R_COMPLEX] = NUMBER
      ATTRIBUTES[RElementTypes.R_INTEGER] = NUMBER

      ATTRIBUTES[RElementTypes.R_LPAR] = PARENTHESES
      ATTRIBUTES[RElementTypes.R_RPAR] = PARENTHESES

      ATTRIBUTES[RElementTypes.R_LBRACE] = BRACES
      ATTRIBUTES[RElementTypes.R_RBRACE] = BRACES

      ATTRIBUTES[RElementTypes.R_LBRACKET] = BRACKETS
      ATTRIBUTES[RElementTypes.R_LDBRACKET] = BRACKETS
      ATTRIBUTES[RElementTypes.R_RBRACKET] = BRACKETS
      ATTRIBUTES[RElementTypes.R_RDBRACKET] = BRACKETS

      ATTRIBUTES[RElementTypes.R_COMMA] = COMMA
      ATTRIBUTES[RElementTypes.R_SEMI] = SEMICOLON

      ATTRIBUTES[RTokenTypes.END_OF_LINE_COMMENT] = LINE_COMMENT
      ATTRIBUTES[RTokenTypes.ROXYGEN_COMMENT] = DOC_COMMENT

      ATTRIBUTES[RTokenTypes.BAD_CHARACTER] = BAD_CHARACTER
    }
  }
}
